File: test/error-bound/trig-hyp.ts
/**
 * @file Pre-test for trigonometric and hyperbolic functions and their inverses
 */

import {
  abs2, add11, gt21,
  sin_1, sin_2, tan_1, tan_2, sec_1, sec_2,
  cos_1, cos_2, cot_1, cot_2, csc_1, csc_2,
  sinh_1, sinh_2, tanh_1, tanh_2, sech_1, sech_2,
  cosh_1, cosh_2, coth_1, coth_2, csch_1, csch_2,
  asin_1, asin_2, atan_1, atan_2, asec_1, asec_2,
  acos_1, acos_2, acot_1, acot_2, acsc_1, acsc_2,
  asinh_1, asinh_2,
  lt21,
} from '../../src/index';

import {
  FnSig, UnionToIntersection, randomFn, collectOutputs, initArgsList, FnBySigOpt,
  ldexp,
  ldexp2,
  // ldexp,
} from '../utils';

import { exponent } from '@lvlte/ulp';
import fs from 'node:fs';

// Functions to test grouped by signature
const fnBySig = {
  'op1': {sin_1, cos_1, tan_1, cot_1, sec_1, csc_1,
          asin_1, acos_1, atan_1, acot_1, asec_1, acsc_1,
          asinh_1},
  'op2': {sin_2, cos_2, tan_2, cot_2, sec_2, csc_2,
          asin_2, acos_2, atan_2, acot_2, asec_2, acsc_2,
          asinh_2},
  // functions defined in terms of e^x are tested apart from op1/op2 group
  'exp1': {sinh_1, cosh_1, tanh_1, coth_1, sech_1, csch_1},
  'exp2': {sinh_2, cosh_2, tanh_2, coth_2, sech_2, csch_2},
} satisfies FnBySigOpt;

type FnBySig = typeof fnBySig;
type TestedFunctions = UnionToIntersection<FnBySig[keyof FnBySig]>;
type FnName = keyof TestedFunctions;
type ArgsListBySig = { [K in keyof FnBySig]: Parameters<FnSig[K]>[] };
type FnOutputList = { [K in FnName]: ReturnType<TestedFunctions[K]>[] };
type FnArgs = { [K in FnName]: Parameters<TestedFunctions[K]> };

/**
 * Return the appropriate processArgs callback for the given function. The role
 * of the callback is to make the function parameters fit its domain if needed.
 */
function processArgsFn(fnName: FnName): Function {
  switch (fnName) {
    case 'asin_1':
    case 'acos_1':
      return (...args: FnArgs[typeof fnName]) => {
        if (Math.abs(args[0]) > 1) {
          const e = exponent(args[0]);
          args[0] = ldexp(args[0], -(1 + e + (e % 2)));
        }
        return args;
      }

    case 'asin_2':
    case 'acos_2':
      return (...args: FnArgs[typeof fnName]) => {
        const x = args[0];
        if (gt21(abs2(x), 1)) {
          const e = exponent(x[0]);
          args[0] = ldexp2(x, -(1 + e + (e % 2)));
        }
        return args;
      }

    case 'tanh_1':
    case 'coth_1':
      // f(x) = 1 + ε with |ε| < u² for |x| > 37.09
      return (...args: FnArgs[typeof fnName]) => (args[0] = args[0] % 37, args);

    case 'tanh_2':
    case 'coth_2':
      return (...args: FnArgs[typeof fnName]) => {
        const [xhi, xlo] = args[0];
        args[0] = add11(xhi % 37, xlo);
        return args;
      }

    case 'asec_1':
    case 'acsc_1':
      return (...args: FnArgs[typeof fnName]) => {
        if (Math.abs(args[0]) < 1) {
          args[0] = ldexp(args[0], -exponent(args[0]));
        }
        return args;
      }

    case 'asec_2':
    case 'acsc_2':
      return (...args: FnArgs[typeof fnName]) => {
        const x = args[0];
        if (lt21(abs2(x), 1)) {
          args[0] = ldexp2(x, -exponent(x[0]));
        }
        return args;
      }
  }

  return (...args: FnArgs[typeof fnName]) => args;
}

// Pseudo-random number generator
const SEED = Math.sqrt(2);
const random = randomFn(SEED, true);

// Lists of arguments (grouped by FnSig) to pass to the TestedFunctions
const argsList = initArgsList<ArgsListBySig>(fnBySig);

const emin = -106;
const emax = 53;

for (let exp = emin; exp <= emax; exp++) {
  for (const sign of [1, -1]) {
    for (let r = 0; r < 100; r++) {
      const x = random(exp, sign);
      const y = random(exp - 52, sign);
      const xy = add11(x, y);
      if ([x, y, ...xy].some(v => !Number.isFinite(v))) {
        continue;
      }
      argsList['op1'].push([x]);
      argsList['op2'].push([xy]);
    }
  }
}

const e_negx = Math.log2(-Math.log(Number.MIN_VALUE)) - 1;
const e_posx = Math.log2(+Math.log(Number.MAX_VALUE)) - 1;
for (const [sign, emax] of [[1, e_posx], [-1, e_negx]]) {
  const emaxint = Math.floor(emax);
  for (let exp = emin; exp <= emax; exp = exp == emaxint ? emax : exp+1) {
    for (let r = 0; r < 200; r++) {
      const x = random(exp, sign);
      const y = random(exp - 52, sign);
      const xy = add11(x, y);
      argsList['exp1'].push([x]);
      argsList['exp2'].push([xy]);
    }
  }
}

const fnOutput = collectOutputs<FnOutputList>(fnBySig, argsList, processArgsFn);
const testset = { argsList, fnOutput };

// Export as JSON
const testsetJSON = JSON.stringify(testset);
fs.writeFileSync('test/error-bound/testset/trig-hyp.json', testsetJSON, 'utf8');

console.log('prerun trig.ts done');

File: test/error-bound/trig-hyp.jl
# Test the accuracy of twofloat's trigonometric/hyperbolic functions
# Function inputs/outputs dataset is created by /test/error-bound/trig-hyp.ts

json = read("$(@__DIR__)/testset/trig-hyp.json", String)
testset = JSON.parse(json, TestSet; null=NaN)

args_list = testset.argsList
fn_output = testset.fnOutput

coverage = OrderedDict(keys(fn_output) .=> false)
overflow = OrderedDict(keys(fn_output) .=> 0)

println()
@testset verbose = true "Trigonometric functions ─" begin ######################

    @testset "sin_1" begin
        _test(Dict(
            "fn" => "sin_1",
            "args" => args_list.op1,
            "rel_err_bound" => big(2.0)^-90,
            "compute" => x -> sin(big(x))
        ))
    end

    @testset "sin_2" begin
        _test(Dict(
            "fn" => "sin_2",
            "args" => args_list.op2,
            "rel_err_bound" => big(2.0)^-90,
            "compute" => ((xhi, xlo),) -> sin(big(xhi) + big(xlo))
        ))
    end

    @testset "cos_1" begin
        _test(Dict(
            "fn" => "cos_1",
            "args" => args_list.op1,
            "rel_err_bound" => big(2.0)^-90,
            "compute" => x -> cos(big(x))
        ))
    end

    @testset "cos_2" begin
        _test(Dict(
            "fn" => "cos_2",
            "args" => args_list.op2,
            "rel_err_bound" => big(2.0)^-90,
            "compute" => ((xhi, xlo),) -> cos(big(xhi) + big(xlo))
        ))
    end

    @testset "tan_1" begin
        _test(Dict(
            "fn" => "tan_1",
            "args" => args_list.op1,
            "rel_err_bound" => big(2.0)^-90,
            "compute" => x -> tan(big(x))
        ))
    end

    @testset "tan_2" begin
        _test(Dict(
            "fn" => "tan_2",
            "args" => args_list.op2,
            "rel_err_bound" => big(2.0)^-90,
            "compute" => ((xhi, xlo),) -> tan(big(xhi) + big(xlo))
        ))
    end

    @testset "cot_1" begin
        _test(Dict(
            "fn" => "cot_1",
            "args" => args_list.op1,
            "rel_err_bound" => big(2.0)^-90,
            "compute" => x -> cot(big(x))
        ))
    end

    @testset "cot_2" begin
        _test(Dict(
            "fn" => "cot_2",
            "args" => args_list.op2,
            "rel_err_bound" => big(2.0)^-90,
            "compute" => ((xhi, xlo),) -> cot(big(xhi) + big(xlo))
        ))
    end

    @testset "sec_1" begin
        _test(Dict(
            "fn" => "sec_1",
            "args" => args_list.op1,
            "rel_err_bound" => big(2.0)^-90,
            "compute" => x -> sec(big(x))
        ))
    end

    @testset "sec_2" begin
        _test(Dict(
            "fn" => "sec_2",
            "args" => args_list.op2,
            "rel_err_bound" => big(2.0)^-90,
            "compute" => ((xhi, xlo),) -> sec(big(xhi) + big(xlo))
        ))
    end

    @testset "csc_1" begin
        _test(Dict(
            "fn" => "csc_1",
            "args" => args_list.op1,
            "rel_err_bound" => big(2.0)^-90,
            "compute" => x -> csc(big(x))
        ))
    end

    @testset "csc_2" begin
        _test(Dict(
            "fn" => "csc_2",
            "args" => args_list.op2,
            "rel_err_bound" => big(2.0)^-90,
            "compute" => ((xhi, xlo),) -> csc(big(xhi) + big(xlo))
        ))
    end
end

println()
@testset verbose = true "Inverse Trigonometric functions ─" begin ######################

    @testset "asin_1" begin
        _test(Dict(
            "fn" => "asin_1",
            "args" => args_list.op1,
            "process_args" => function(x)
                abs(x) <= 1 && return (x,)
                e = exponent(x)
                p = -(1 + e + (e % 2))
                return (x*2.0^p,)
            end,
            "compute" => x -> asin(big(x))
        ))
    end

    @testset "asin_2" begin
        _test(Dict(
            "fn" => "asin_2",
            "args" => args_list.op2,
            "process_args" => function((xhi, xlo),)
                x = big(xhi) + big(xlo)
                abs(x) <= 1 && return ((xhi, xlo),)
                e = exponent(x)
                p = -(1 + e + (e % 2))
                return ((xhi*2.0^p, xlo*2.0^p),)
            end,
            "compute" => ((xhi, xlo),) -> asin(big(xhi) + big(xlo))
        ))
    end

    @testset "acos_1" begin
        _test(Dict(
            "fn" => "acos_1",
            "args" => args_list.op1,
            "rel_err_bound" => big(2.0)^-95,
            "process_args" => function(x)
                abs(x) <= 1 && return (x,)
                e = exponent(x)
                p = -(1 + e + (e % 2))
                return (x*2.0^p,)
            end,
            "compute" => x -> acos(big(x))
        ))
    end

    @testset "acos_2" begin
        _test(Dict(
            "fn" => "acos_2",
            "args" => args_list.op2,
            "rel_err_bound" => big(2.0)^-95,
            "process_args" => function((xhi, xlo),)
                x = big(xhi) + big(xlo)
                abs(x) <= 1 && return ((xhi, xlo),)
                e = exponent(x)
                p = -(1 + e + (e % 2))
                return ((xhi*2.0^p, xlo*2.0^p),)
            end,
            "compute" => ((xhi, xlo),) -> acos(big(xhi) + big(xlo))
        ))
    end

    @testset "atan_1" begin
        _test(Dict(
            "fn" => "atan_1",
            "args" => args_list.op1,
            "compute" => x -> atan(big(x))
        ))
    end

    @testset "atan_2" begin
        _test(Dict(
            "fn" => "atan_2",
            "args" => args_list.op2,
            "compute" => ((xhi, xlo),) -> atan(big(xhi) + big(xlo))
        ))
    end

    @testset "acot_1" begin
        _test(Dict(
            "fn" => "acot_1",
            "args" => args_list.op1,
            "compute" => x -> acot(big(x))
        ))
    end

    @testset "acot_2" begin
        _test(Dict(
            "fn" => "acot_2",
            "args" => args_list.op2,
            "compute" => ((xhi, xlo),) -> acot(big(xhi) + big(xlo))
        ))
    end

    @testset "asec_1" begin
        _test(Dict(
            "fn" => "asec_1",
            "args" => args_list.op1,
            "rel_err_bound" => big(2.0)^-90,
            "process_args" => function(x)
                abs(x) ≥ 1 && return (x,)
                return (ldexp(x, -exponent(x)),)
            end,
            "compute" => x -> asec(big(x))
        ))
    end

    @testset "asec_2" begin
        _test(Dict(
            "fn" => "asec_2",
            "args" => args_list.op2,
            "rel_err_bound" => big(2.0)^-90,
            "process_args" => function((xhi, xlo),)
                abs(big(xhi) + big(xlo)) ≥ 1 && return ((xhi, xlo),)
                p = -exponent(xhi)
                return ((ldexp(xhi, p), ldexp(xlo, p)),)
            end,
            "compute" => ((xhi, xlo),) -> asec(big(xhi) + big(xlo))
        ))
    end
end

println()
@testset verbose = true "Hyperbolic functions ────" begin ######################

    @testset "sinh_1" begin
        _test(Dict(
            "fn" => "sinh_1",
            "args" => args_list.exp1,
            "rel_err_bound_uf" => big(2.0)^-75,
            "compute" => x -> sinh(big(x)),
            "underflow" => (r, (x,)) -> abs(u^2 * exp(big(x))) < ε₀
        ))
    end

    @testset "sinh_2" begin
        _test(Dict(
            "fn" => "sinh_2",
            "args" => args_list.exp2,
            "rel_err_bound_uf" => big(2.0)^-75,
            "compute" => ((xhi, xlo),) -> sinh(big(xhi) + big(xlo)),
            "underflow" => (r, ((xhi, xlo),)) -> abs(u^2 * exp(big(xhi) + big(xlo))) < ε₀
        ))
    end

    @testset "cosh_1" begin
        _test(Dict(
            "fn" => "cosh_1",
            "args" => args_list.exp1,
            "rel_err_bound_uf" => big(2.0)^-75,
            "compute" => x -> cosh(big(x)),
            "underflow" => (r, (x,)) -> abs(u^2 * exp(big(x))) < ε₀
        ))
    end

    @testset "cosh_2" begin
        _test(Dict(
            "fn" => "cosh_2",
            "args" => args_list.exp2,
            "rel_err_bound_uf" => big(2.0)^-75,
            "compute" => ((xhi, xlo),) -> cosh(big(xhi) + big(xlo)),
            "underflow" => (r, ((xhi, xlo),)) -> abs(u^2 * exp(big(xhi) + big(xlo))) < ε₀
        ))
    end

    @testset "tanh_1" begin
        _test(Dict(
            "fn" => "tanh_1",
            "args" => args_list.exp1,
            # "rel_err_bound" => big(2.0)^-95,
            "process_args" => x -> x % 37,
            "compute" => x -> tanh(big(x))
        ))
    end

    @testset "tanh_2" begin
        _test(Dict(
            "fn" => "tanh_2",
            "args" => args_list.exp2,
            "process_args" => ((xhi, xlo),) -> (twosum(xhi % 37, xlo),),
            "compute" => ((xhi, xlo),) -> tanh(big(xhi) + big(xlo))
        ))
    end

    @testset "coth_1" begin
        _test(Dict(
            "fn" => "coth_1",
            "args" => args_list.exp1,
            "process_args" => x -> x % 37,
            "compute" => x -> coth(big(x)),
            "underflow" => (r, (x,)) -> abs(u^2 * exp(big(x))) < ε₀
        ))
    end

    @testset "coth_2" begin
        _test(Dict(
            "fn" => "coth_2",
            "args" => args_list.exp2,
            "process_args" => ((xhi, xlo),) -> (twosum(xhi % 37, xlo),),
            "compute" => ((xhi, xlo),) -> coth(big(xhi) + big(xlo)),
            "underflow" => (r, ((xhi, xlo),)) -> abs(u^2 * exp(big(xhi) + big(xlo))) < ε₀
        ))
    end

    @testset "sech_1" begin
        _test(Dict(
            "fn" => "sech_1",
            "args" => args_list.exp1,
            "compute" => x -> sech(big(x))
        ))
    end

    @testset "sech_2" begin
        _test(Dict(
            "fn" => "sech_2",
            "args" => args_list.exp2,
            "compute" => ((xhi, xlo),) -> sech(big(xhi) + big(xlo))
        ))
    end

    @testset "csch_1" begin
        _test(Dict(
            "fn" => "csch_1",
            "args" => args_list.exp1,
            "compute" => x -> csch(big(x))
        ))
    end

    @testset "csch_2" begin
        _test(Dict(
            "fn" => "csch_2",
            "args" => args_list.exp2,
            "compute" => ((xhi, xlo),) -> csch(big(xhi) + big(xlo))
        ))
    end
end

##

println()
@testset "Trig functions coverage ─" begin
    for (fn, covered) in coverage
        @test (fn, covered) == (fn, true)
    end
end

overflowed = filter(((fn , ov_count),) -> ov_count > 0, overflow)
if !isempty(overflowed)
    println()
    @info ["overflow\n ", (rpad(k, 20, ' ') * "$v\n " for (k, v) in overflowed)...] |> join
end
